lib/llvmopencl/FlattenBarrierSubs.cc

```diff
@@ -0,0 +1,127 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <iostream>
+#include <string>
+
+#include "CompilerWarnings.h"
+IGNORE_COMPILER_WARNING("-Wunused-parameter")
+
+#include "config.h"
+#include "pocl.h"
+
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+#include <llvm/IR/Instructions.h>
+
+#include "Workgroup.h"
+
+POP_COMPILER_DIAGS
+
+using namespace llvm;
+
+namespace {
+class FlattenBarrierSubs : public ModulePass {
+
+public:
+  static char ID;
+  FlattenBarrierSubs() : ModulePass(ID) {}
+
+  virtual bool runOnModule(Module &M);
+};
+
+} // namespace
+
+//#define DEBUG_FLATTEN
+
+extern cl::opt<std::string> KernelName;
+
+char FlattenBarrierSubs::ID = 0;
+static RegisterPass<FlattenBarrierSubs>
+    X("flatten-barrier-subs",
+      "Kernel function flattening pass - flatten subroutines with barriers "
+      "and/or local arguments");
+
+static bool recursivelyInlineBarrierUsers(Function *F, bool changeInlineFlag) {
+
+  bool callsBarrier = false;
+
+#ifdef DEBUG_FLATTEN
+  std::cerr << "### FlattenBarrierSubs: SCANNING " << F->getName().str()
+            << std::endl;
+#endif
+
+  for (Function::iterator I = F->begin(), E = F->end(); I != E; ++I) {
+    for (BasicBlock::iterator BI = I->begin(), BE = I->end(); BI != BE; ++BI) {
+      Instruction *Instr = dyn_cast<Instruction>(BI);
+      if (!llvm::isa<CallInst>(Instr))
+        continue;
+      CallInst *CallInstr = dyn_cast<CallInst>(Instr);
+      Function *callee = CallInstr->getCalledFunction();
+
+      if ((callee == nullptr) ||  callee->getName().startswith("llvm."))
+        continue;
+
+      if (callee->getName().equals("pocl.barrier") ||
+          callee->getName().equals("_Z7barrierj"))
+        callsBarrier = true;
+      // we cannot break the loop here, since we want to inline all functions
+      // that could call a barrier, not just the first one
+      else if (recursivelyInlineBarrierUsers(callee, true))
+        callsBarrier = true;
+    }
+  }
+
+  if (changeInlineFlag & callsBarrier) {
+    F->addFnAttr(Attribute::AlwaysInline);
+    F->setLinkage(llvm::GlobalValue::InternalLinkage);
+#ifdef DEBUG_FLATTEN
+    std::cerr << "### FlattenBarrierSubs: AlwaysInline ENABLED on "
+              << F->getName().str() << std::endl;
+#endif
+  }
+  return callsBarrier;
+}
+
+bool FlattenBarrierSubs::runOnModule(Module &M) {
+  bool changed = false;
+  for (llvm::Module::iterator i = M.begin(), e = M.end(); i != e; ++i) {
+    llvm::Function *F = &*i;
+    if (F->isDeclaration())
+      continue;
+
+    if (KernelName == F->getName() ||
+        (KernelName == "" && pocl::Workgroup::isKernelToProcess(*F))) {
```

Comment:
Why the empty name check?

Reply:
Nope, no clue at all, cannot remember :)

Reply:
Sorry, i suspect you don't remember. I'll take a look if removing it breaks anything.

Reply:
OK, i believe the entire second check is now redundant, as the LLVM passes now run on a Module with just one kernel present.